crates/aptos-jwk-consensus/src/jwk_observer.rs

```diff
@@ -0,0 +1,136 @@
+    issuer: String,
+    jwks_uri: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct JWKsResponse {
+    keys: Vec<serde_json::Value>,
+}
+
+/// Given an Open ID configuration URL, fetch its JWKs.
+pub async fn fetch_jwks(my_addr: AccountAddress, config_url: Vec<u8>) -> Result<Vec<JWK>> {
+    if cfg!(feature = "smoke-test") {
+        use reqwest::header;
+        let maybe_url = String::from_utf8(config_url);
+        let jwk_url = maybe_url?;
+        let client = reqwest::Client::new();
+        let JWKsResponse { keys } = client
+            .get(jwk_url.as_str())
+            .header(header::COOKIE, my_addr.to_hex())
+            .send()
+            .await?
+            .json()
+            .await?;
+        let jwks = keys.into_iter().map(JWK::from).collect();
+        Ok(jwks)
+    } else {
+        let maybe_url = String::from_utf8(config_url);
+        let config_url = maybe_url?;
+        let client = reqwest::Client::new();
+        let OpenIDConfiguration { jwks_uri, .. } =
+            client.get(config_url.as_str()).send().await?.json().await?;
+        let JWKsResponse { keys } = client.get(jwks_uri.as_str()).send().await?.json().await?;
+        let jwks = keys.into_iter().map(JWK::from).collect();
+        Ok(jwks)
+    }
+}
+
+/// A process thread that periodically fetch JWKs of a provider and push it back to JWKManager.
+pub struct JWKObserver {
+    close_tx: oneshot::Sender<()>,
+    join_handle: JoinHandle<()>,
+}
+
+impl JWKObserver {
+    pub fn spawn(
+        my_addr: AccountAddress,
+        issuer: Issuer,
+        config_url: Vec<u8>,
+        fetch_interval: Duration,
+        observation_tx: aptos_channel::Sender<(), (Issuer, Vec<JWK>)>,
+    ) -> Self {
+        let (close_tx, close_rx) = oneshot::channel();
+        let join_handle = tokio::spawn(Self::thread_main(
+            fetch_interval,
+            my_addr,
+            issuer.clone(),
+            config_url.clone(),
+            observation_tx,
+            close_rx,
+        ));
+        info!(
+            "[JWK] observer spawned, issuer={:?}, config_url={:?}",
+            String::from_utf8(issuer),
+            String::from_utf8(config_url)
+        );
+        Self {
+            close_tx,
+            join_handle,
+        }
+    }
+
+    async fn thread_main(
+        fetch_interval: Duration,
+        my_addr: AccountAddress,
+        issuer: Issuer,
+        open_id_config_url: Vec<u8>,
+        observation_tx: aptos_channel::Sender<(), (Issuer, Vec<JWK>)>,
+        close_rx: oneshot::Receiver<()>,
+    ) {
+        let mut interval = tokio::time::interval(fetch_interval);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let mut close_rx = close_rx.into_stream();
+        loop {
+            tokio::select! {
+                _ = interval.tick().fuse() => {
+                    let result = fetch_jwks(my_addr, open_id_config_url.clone()).await;
+                    debug!("observe_result={:?}", result);
```

Comment:
print more information? like version, issuer, my_id?

Reply:
will have a new PR to cover log + metrics